### PR TITLE
Feat : Back up - Part 3: Add database reader which reads all items at once

### PR DESCRIPTION
**Pull request by gizemb**
## What's new in this PR?

Big tables might require [reading as batches](2946), but for small database tables, we can read all data at once for convenience. This PR adds a BackUpIOHandler which reads all entities from a database at once.

### Testing

Unit tests are added.

Note: I'm open for better naming suggestions

### CODE REVIEW

**Comment by gizemb** [comment]
build again

File: app/src/main/kotlin/com/waz/zclient/feature/backup/io/database/SingleReadDatabaseIOHandler.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    override fun readIterator() = object : Iterator<T> {
+        val listIterator by lazy { singleReadDao.getAll().iterator() }
+
+        override fun hasNext(): Boolean = listIterator.hasNext()
+
+        override fun next(): T = listIterator.next()
+    }
+}
+
+interface SingleReadDao<T> {
```
Comment (BradleyWire) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/io/database/SingleReadDatabaseIOHandler.kt:23]: So the requirements for a DAO to extend this interface is that it doesn't have a dynamic size of rows? i.e. KeyValues would be a SingleRead but Messages clearly wouldn't be?

**Comment by gizemb** [comment]
build again

File: app/src/main/kotlin/com/waz/zclient/feature/backup/io/database/SingleReadDatabaseIOHandler.kt (hunk 1 above)
Comment (gizemb) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/io/database/SingleReadDatabaseIOHandler.kt:23]: yes, that's what i thought

**Comment by gizemb** [comment]
build again

**Comment by gizemb** [comment]
build again

**Comment by gizemb** [comment]
build again

**Pull request merged**